### PR TITLE
TranscoderPool implemented as sorted linked list with optional hints

### PR DESCRIPTION
**Pull request by yondonfu**
Changed the TranscoderPool implementation to use a sorted linked list with optional hints instead of a combination of a min-heap and max-heap.

TranscoderPool is a sorted doubly linked list
- It accepts an optional hint in the form of `(worseTranscoder, betterTranscoder)` where worseTranscoder is a transcoder in the pool with stake less than or equal to the new transcoder and betterTranscoder is a transcoder in the pool with stake greater than or equal to the new transcoder
- Clients interacting with the contracts could calculate the hint off-chain
- If the hint is a pair of correct bounding transcoders in the list, insertion is O(1)
- If the hint is not a pair of correct bounding transcoders, either because of a change in the list or because the hint is just wrong, we ascend/descend the list to look for the appropriate insertion point using the hint as a starting point if applicable
- If no hint is provided, we just ascend the list starting from the head to look for an insertion point
- Using hints, insertion is O(1) reads/writes in the best case (if the hint is a pair of correct bounding transcoders) and O(n) reads + O(1) writes in the worst case, but in practice the number of reads will not be as large as n
- The number of reads in the worst case will be dependent on the frequency of bonding - if bonding occurs frequently there will be more reads, if bonding occurs infrequently there will be less reads
- I've included some unit tests, but still need to cover all the edge cases

The BondingManager currently does not provide hints when adding new transcoders or updating the stakes of existing transcoders as this requires a change in the function signatures for `bond`, `transcoder`, `unbond`, `reward` - this change is left out of this PR.

### CODE REVIEW

File: contracts/bonding/libraries/TranscoderPool.sol [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+            addTranscoder(self, _transcoder, newStake, _worseTranscoder, _betterTranscoder);
+        }
+    }
+
+    /*
+     * @dev Checks if the list contains a transcoder
+     * @param _transcoder Address of transcoder
+     */
+    function contains(Data storage self, address _transcoder) public view returns (bool) {
+        return self.head == _transcoder || self.tail == _transcoder || self.nodes[_transcoder].prevTranscoder != address(0) || self.nodes[_transcoder].nextTranscoder != address(0);
```
Comment (dob) [inline contracts/bonding/libraries/TranscoderPool.sol]: Can we just check the mapping to see if the returned node for this transcoder address has non-zero stake?

File: contracts/bonding/libraries/TranscoderPool.sol [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+     */
+    function getBestTranscoder(Data storage self) public view returns (address) {
+        return self.tail;
+    }
+
+    /*
+     * @dev Returns the previous transcoder in the list for a transcoder
+     * @param _transcoder Address of transcoder
+     */
+    function getNextTranscoder(Data storage self, address _transcoder) public view returns (address) {
```
Comment (dob) [inline contracts/bonding/libraries/TranscoderPool.sol]: `getNext` and `getPrev` are very confusing, especially since the comments mix prev + next.

File: contracts/bonding/libraries/TranscoderPool.sol [hunk 3]
```diff
@@ -0,0 +1,372 @@
+pragma solidity ^0.4.17;
+
+import "zeppelin-solidity/contracts/math/SafeMath.sol";
+
+
+/*
+ * @title A pool of registered transcoders and their stakes implemented as a sorted doubly linked list
+ */
+library TranscoderPool {
```
Comment (dob) [inline contracts/bonding/libraries/TranscoderPool.sol]: General comment about the transcoder pool is that I think that the implementation is a little bit verbose and potentially confusing because we introduce transcoder specific terminology. In reality however, are we doing anything at all that is unique to transcoders...or is it just a sorted doubly linked list data structure, using a node's value as the sort key? With the ability for the node's value to be updated triggering a resort?

If this is actually the case, the implementation can potentially avoid all the `better`, `worse`, `transcoder` sepcific terminology, and just use `next`, `prev`, and update `value` functions?

Open to discussion on this. It may be more auditable, reusable, and convenient.

File: contracts/bonding/libraries/TranscoderPool.sol [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+        }
+
+        self.nodes[_transcoder].stake = _stake;
+
+        if (worseTranscoder == address(0) && betterTranscoder == address(0)) {
+            // Insert as head and tail
+            self.head = _transcoder;
+            self.tail = _transcoder;
+        } else if (worseTranscoder == address(0)) {
+            // Insert as head
```
[the remark this answers is not in the document]
Comment (yondonfu) [reply contracts/bonding/libraries/TranscoderPool.sol]: The idea is that if `findBoundingTranscoders` returns a `(worseTranscoder, betterTranscoder)` such that `worseTranscoder` is null, then `betterTranscoder` is the head and there is no transcoder with less stake than the new transcoder so the new transcoder should become the new head. The implementation could probably be clearer...will work on that

File: contracts/bonding/libraries/TranscoderPool.sol [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+            self.nodes[_transcoder].nextTranscoder = self.head;
+            self.nodes[self.head].prevTranscoder = _transcoder;
+            self.head = _transcoder;
+        } else if (betterTranscoder == address(0)) {
+            // Insert as tail
```
[the remark this answers is not in the document]
Comment (yondonfu) [reply contracts/bonding/libraries/TranscoderPool.sol]: Same comment from above for this except that `worseTranscoder` is the tail and there is no transcoder with more stake than the new transcoder so the new transcoder should become the tail

**Comment by yondonfu** [comment]
> Does the fact that every single accounting update to bonding amounts now need to potentially update the sorting of the list make all these actions significantly more expensive?

If no hints are provided, it is more expensive since there are O(n) reads to find the insertion point. If hints are provided, it is O(1) reads in the best case (if the hint is the correct insertion point) and O(n) reads in the worst case where the number of reads in practice is likely smaller than n. So on average, if hints are provided, I think the sorted doubly linked list is less expensive than the alternative implementation with a min-heap for the candidate pool and an unsorted linked list with max/min pointers for the reserve pool.

I agree that this would all be clearer if it were implemented as a generic data structure rather than trying to tailor the terminology used specifically for transcoders!

**Comment by yondonfu** [comment]
Pulled out `SortedDoublyLL` into its own library. Also did some refactoring so that the list is sorted in descending order such that the head is the node with the largest key and the tail is the node with the smallest key. As a result, when using the list in the BondingManager we can pull the transcoder with the least stake with `getLast` and pull the top `N` transcoders for the active set by starting with the transcoder with the most stake using `getFirst` and then iterating through the rest using `getNext`

File: contracts/bonding/BondingManager.sol [hunk 6]
```diff
@@ -295,38 +301,36 @@ contract BondingManager is ManagerProxyTarget, IBondingManager {
...
+            t.feeShare = feeShare;
+            t.pricePerSegment = pricePerSegment;
+            // Initialize token pool
+            t.tokenPoolsPerRound[currentRound].init(stake, blockRewardCut, feeShare);
+
+            totalStake = totalStake.add(stake);
 
-            stake = stake.add(transcoderTotalStake(activeTranscoder));
+            // Get next transcoder in the pool
+            currentTranscoder = transcoderPool.getPrevTranscoder(currentTranscoder);
```
[the remark this answers is not in the document]
Comment (yondonfu) [reply contracts/bonding/BondingManager.sol]: Changed so it now uses `getNext` to iterate through the list

File: contracts/bonding/libraries/TranscoderPool.sol [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+
+    // Represents a transcoder in the list
+    struct Node {
+        uint256 stake;                   // Stake for the transcoder represented by this node
+        address nextTranscoder;          // Address of next transcoder in the list
+        address prevTranscoder;          // Address of previous transcoder in the list
+    }
+
+    // Represents a sorted doubly linked list
+    struct Data {
```
[the remark this answers is not in the document]
Comment (yondonfu) [reply contracts/bonding/libraries/TranscoderPool.sol]: Since we're usually using the library by attaching its functions to a struct via the `using A for B` directive I didn't want to shadow the library name by naming the struct the same thing (which the compiler gives you a warning for when the struct is defined within the library). I decided to just give the struct a generic name like `Data` since it'll be referenced as `SortedDoublyLL.Data` in any contract that is using the `SortedDoublyLL` library. I'm definitely not attached to the naming though and am open to better suggestions!

File: contracts/libraries/SortedDoublyLL.sol [hunk 8]
```diff
@@ -0,0 +1,359 @@
...
+    }
+
+    /*
+     * @dev Increase the key of a node in the list
+     * @param _id Node's id
+     * @param _amount Amount to increase node's key
+     * @param _prevId Id of previous node for the new insert position
+     * @param _nextId Id of next node for the new insert position
+     */
+    function increaseKey(Data storage self, address _id, uint256 _amount, address _prevId, address _nextId) public {
```
Comment (dob) [inline contracts/libraries/SortedDoublyLL.sol]: I was just thinking that maybe the user of the DoublyLL shouldn't have to know whether they are increasing or decreasing the key, they should just call `updateKey()`? What do you think?

**Comment by dob** [comment]
This feels a lot clearer to me now. Nice work. I could see a future implementation where there's still a `TranscoderPools` abstraction in front of the doublyLL that wraps it, such that it could be swapped out easily for a different implementation in the future, but for now I think this is a good straightforward compromise. 

What does it mean that all the methods on the library are public? Can anyone invoke them...or can they only be invoked on the object itself which is only modifiable within other smart contract functions?

One more question...it looks like `maxSize` is settable...but it doesn't move the head/tail pointers around in the case of reducing the max size? Should it?

**Comment by yondonfu** [comment]
> or can they only be invoked on the object itself which is only modifiable within other smart contract functions?

Yep, the library itself has no state

> it looks like maxSize is settable...but it doesn't move the head/tail pointers around in the case of reducing the max size? Should it?

So this can be a little tricky if we reduce the max size such that the new size is less than the current size and the tail needs to be updated. We can:

1.  Delete every node after the new tail. We would need to be careful here not to hit the block gas limit - even though each storage delete reimburses the caller, the reimbursement only happens at the end of a call and each storage element deletion costs 5000 gas
2. "Soft delete" every node after the new tail by setting an `exists` boolean to false. The same concerns for 1. apply here except we only have one storage update as opposed to multiple storage deletions and without the gas reimbursement at the end
3. Don't delete nodes after the new tail. The `contains` function would need to be updated because nodes not in the list could still have non-zero keys. One option could be to insert each node with an increasing nonce and a node exists in the list if its nonce is less than the tail's nonce
4. Disallow decreasing the max size of the list

From a cost and simplicity point of view I'm interested in 4. I'm not sure if there is a particular scenario where we would want to decrease the total number of transcoders that are allowed to be registered (the number of _active_ transcoders can be tweaked since we pull the top `N` transcoders from the list at the beginning of a round), but let me know if there is

**Comment by dob** [comment]
Got it, so there's no need to decrease `maxSize`, and we can always move the number of active transcoders just by changing `N`. So option 4 seems reasonable.
